### PR TITLE
Update opentelemetry & related dependencies.

### PR DESCRIPTION
**Pull request by branlwyd**
This required quite a few changes -- this is a major-version update & there were several non-trivial, non-backwards-compatible changes.

Unfortunately, due to the reliance on Tonic, building Janus now requires a protoc compiler when janus_server's "otlp" feature is enabled.

### CODE REVIEW

**Comment by branlwyd** [comment]
(I'm fairly unconfident of some of the changes in `metrics.rs` -- some of the previously-specified options appear to be just gone, rather than configured in a different way, so I made do as best I could. Is there a good way to test these changes out?)

File: janus_server/src/aggregator/aggregation_job_driver.rs [hunk 1]
```diff
@@ -47,24 +50,31 @@ use tracing::{info, warn};
...
+            .init();
         let job_cancel_counter = meter
             .u64_counter("janus_job_cancellations")
             .with_description("Count of cancelled jobs.")
-            .init()
-            .bind(&[]);
-        job_cancel_counter.add(0);
+            .init();
+
+        aggregate_step_failure_counter.add(&Context::current(), 0, &[]);
```
Comment (divergentdave) [inline janus_server/src/aggregator/aggregation_job_driver.rs:72]: As above with `janus_database_transactions_total`, we won't use `janus_step_failures` with an empty set of labels at runtime, so we shouldn't pre-create a record for it with no labels. Instead, we should extract the twelve `add(0)` calls that were previously part of `AggregateStepFailureCounters::new()` into a free function, and call that here and in aggregator.rs.

**Review by divergentdave — CHANGES_REQUESTED** [review]
Looks good, I noted a few small issues below.

I confirmed that there's a hard requirement on protoc for now (though there is an [open issue to revisit this](https://github.com/open-telemetry/opentelemetry-rust/issues/873)) and bound instruments are no longer available via any public API. I didn't find mention of bound instruments in the changelogs, but I presume they were taken out for API simplicity, or because they conflicted with the new `Context` API concept.

File: janus_server/src/metrics.rs [hunk 2]
```diff
@@ -1,11 +1,25 @@
 //! Collection and exporting of application-level metrics for Janus.
 
 use opentelemetry::{
-    metrics::Descriptor,
-    sdk::export::metrics::{Aggregator, AggregatorSelector},
+    runtime::Tokio,
```
[the remark this answers is not in the document]
Comment (branlwyd) [reply janus_server/src/metrics.rs:4]: I had to introduce a third conditionally-compiled `use` stanza for dependencies used by both the prometheus & OTLP features -- Rust doesn't let you import the same name twice, even if the name refers to the same item both times.

If complexity continues to grow, we might want to take another strategy like moving the `use` statements into the conditionally-compiled blocks that use them. That might make things more manageable as long as it doesn't result in too much repitition of imports.

**Review by tgeoghegan — APPROVED** [review]
LGTM, though I agree the changes are pretty extensive so it's hard to have confidence here. However I don't see any harm in deploying this to interop-00 and then just seeing if our dashboards still look right.

**Review by divergentdave — APPROVED** [review]

**Pull request merged**